For a typo `let_some_variable =`, suggest `let some_variable =` instead of `let let_some_variable =`

### Code

```Rust
fn main() {
    let_some_variable = 6;
}
```


### Current output

```Shell
error[E0425]: cannot find value `let_some_variable` in this scope
 --> src/main.rs:2:5
  |
2 |     let_some_variable = 6;
  |     ^^^^^^^^^^^^^^^^^
  |
help: you might have meant to introduce a new binding
  |
2 |     let let_some_variable = 6;
  |     +++
```


### Desired output

```Shell
error[E0425]: cannot find value `let_some_variable` in this scope
 --> src/main.rs:2:5
  |
2 |     let_some_variable = 6;
  |     ^^^^^^^^^^^^^^^^^
  |
help: you might have meant to introduce a new binding
  |
2 |     let some_variable = 6;
  |        +
```


### Rationale and extra context

I accidentally typoed an underscore instead of a space between `let` and the name of the new variable I was trying to declare. If a binding isn't found, and it starts with `let_` (or frankly `letsome_variable`, if you forget/miss the space completely), it seems like it would be more likely to suggest `let ` (`let` plus a space) rather than the current suggestion of an additional `let `.

### Other cases

_No response_

### Anything else?

[@estebank said I should file this ❤️ ](https://hachyderm.io/@ekuber/111405360507141151)

<!-- TRIAGEBOT_START -->

<!-- TRIAGEBOT_ASSIGN_START -->

<!-- TRIAGEBOT_ASSIGN_DATA_START$${"user":"Young-Flash"}$$TRIAGEBOT_ASSIGN_DATA_END -->

<!-- TRIAGEBOT_ASSIGN_END -->
<!-- TRIAGEBOT_END -->

--- a/compiler/rustc_resolve/src/late/diagnostics.rs
+++ b/compiler/rustc_resolve/src/late/diagnostics.rs
@@ -1966,13 +1966,22 @@ impl<'a: 'ast, 'ast, 'tcx> LateResolutionVisitor<'a, '_, 'ast, 'tcx> {
     fn let_binding_suggestion(&mut self, err: &mut Diagnostic, ident_span: Span) -> bool {
         if let Some(Expr { kind: ExprKind::Assign(lhs, ..), .. }) =
             self.diagnostic_metadata.in_assignment
-            && let ast::ExprKind::Path(None, _) = lhs.kind
+            && let ast::ExprKind::Path(None, ref path) = lhs.kind
         {
             if !ident_span.from_expansion() {
+                let (span, text) = match path.segments.first() {
+                    Some(seg) if let Some(name) = seg.ident.as_str().strip_prefix("let") => {
+                        // a special case for #117894
+                        let name = name.strip_prefix("_").unwrap_or(name);
+                        (ident_span, format!("let {name}"))
+                    }
+                    _ => (ident_span.shrink_to_lo(), "let ".to_string()),
+                };
+
                 err.span_suggestion_verbose(
-                    ident_span.shrink_to_lo(),
+                    span,
                     "you might have meant to introduce a new binding",
-                    "let ".to_string(),
+                    text,
                     Applicability::MaybeIncorrect,
                 );
                 return true;

--- a/tests/ui/suggestions/suggest-let-for-assignment.fixed
+++ b/tests/ui/suggestions/suggest-let-for-assignment.fixed
@@ -7,6 +7,12 @@ fn main() {
     let x = "x"; //~ ERROR cannot find value `x` in this scope
     println!("x: {}", x); //~ ERROR cannot find value `x` in this scope
 
+    let some_variable = 6; //~ cannot find value `let_some_variable` in this scope
+    println!("some_variable: {}", some_variable); //~ ERROR cannot find value `some_variable` in this scope
+
+    let other_variable = 6; //~ cannot find value `letother_variable` in this scope
+    println!("other_variable: {}", other_variable); //~ ERROR cannot find value `other_variable` in this scope
+
     if x == "x" {
         //~^ ERROR cannot find value `x` in this scope
         println!("x is 1");

--- a/tests/ui/suggestions/suggest-let-for-assignment.rs
+++ b/tests/ui/suggestions/suggest-let-for-assignment.rs
@@ -7,6 +7,12 @@ fn main() {
     x = "x"; //~ ERROR cannot find value `x` in this scope
     println!("x: {}", x); //~ ERROR cannot find value `x` in this scope
 
+    let_some_variable = 6; //~ cannot find value `let_some_variable` in this scope
+    println!("some_variable: {}", some_variable); //~ ERROR cannot find value `some_variable` in this scope
+
+    letother_variable = 6; //~ cannot find value `letother_variable` in this scope
+    println!("other_variable: {}", other_variable); //~ ERROR cannot find value `other_variable` in this scope
+
     if x == "x" {
         //~^ ERROR cannot find value `x` in this scope
         println!("x is 1");

--- a/tests/ui/suggestions/suggest-let-for-assignment.stderr
+++ b/tests/ui/suggestions/suggest-let-for-assignment.stderr
@@ -32,14 +32,48 @@ error[E0425]: cannot find value `x` in this scope
 LL |     println!("x: {}", x);
    |                       ^ not found in this scope
 
+error[E0425]: cannot find value `let_some_variable` in this scope
+  --> $DIR/suggest-let-for-assignment.rs:10:5
+   |
+LL |     let_some_variable = 6;
+   |     ^^^^^^^^^^^^^^^^^
+   |
+help: you might have meant to introduce a new binding
+   |
+LL |     let some_variable = 6;
+   |     ~~~~~~~~~~~~~~~~~
+
+error[E0425]: cannot find value `some_variable` in this scope
+  --> $DIR/suggest-let-for-assignment.rs:11:35
+   |
+LL |     println!("some_variable: {}", some_variable);
+   |                                   ^^^^^^^^^^^^^ not found in this scope
+
+error[E0425]: cannot find value `letother_variable` in this scope
+  --> $DIR/suggest-let-for-assignment.rs:13:5
+   |
+LL |     letother_variable = 6;
+   |     ^^^^^^^^^^^^^^^^^
+   |
+help: you might have meant to introduce a new binding
+   |
+LL |     let other_variable = 6;
+   |     ~~~~~~~~~~~~~~~~~~
+
+error[E0425]: cannot find value `other_variable` in this scope
+  --> $DIR/suggest-let-for-assignment.rs:14:36
+   |
+LL |     println!("other_variable: {}", other_variable);
+   |                                    ^^^^^^^^^^^^^^ not found in this scope
+
 error[E0425]: cannot find value `x` in this scope
-  --> $DIR/suggest-let-for-assignment.rs:10:8
+  --> $DIR/suggest-let-for-assignment.rs:16:8
    |
 LL |     if x == "x" {
    |        ^ not found in this scope
 
 error[E0425]: cannot find value `y` in this scope
-  --> $DIR/suggest-let-for-assignment.rs:15:5
+  --> $DIR/suggest-let-for-assignment.rs:21:5
    |
 LL |     y = 1 + 2;
    |     ^
@@ -50,11 +84,11 @@ LL |     let y = 1 + 2;
    |     +++
 
 error[E0425]: cannot find value `y` in this scope
-  --> $DIR/suggest-let-for-assignment.rs:16:23
+  --> $DIR/suggest-let-for-assignment.rs:22:23
    |
 LL |     println!("y: {}", y);
    |                       ^ not found in this scope
 
-error: aborting due to 7 previous errors
+error: aborting due to 11 previous errors
 
 For more information about this error, try `rustc --explain E0425`.

